### PR TITLE
[16.0][MIG] account_customer_wallet

### PR DESCRIPTION
**Pull request by carmenbianca**
## Odoo task (if applicable)

https://gestion.coopiteasy.be/web#id=10368&action=479&model=project.task&view_type=form&menu_id=536

## Checklist before approval

- [ ] Tests are present (or not needed).
- [ ] Credits/copyright have been changed correctly.
- [ ] Change log snippet is present.
- [ ] (If a new module) Moving this to OCA has been considered.

### CODE REVIEW

File: account_customer_wallet/tests/common.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                active_model="account.move",
+                active_ids=invoice.ids,
+            )
+            .create(
+                {
+                    "journal_id": self.customer_wallet_journal.id,
+                }
+            )
+        )
+        return register._create_payments()
```
Comment (carmenbianca) [inline account_customer_wallet/tests/common.py:127]: So this kind of works, and triggers the balance check in `account_payment.py` correctly, but it does NOT create `account.move.line`s that affect the customer account wallet. I do not know which code to alter where to create an invoice OR a payment register that credits the wallet account.

File: account_customer_wallet/models/account_payment.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+            lambda x: x.journal_id.is_customer_wallet_journal
+        )
+
+        for payment in wallet_payments:
+            if (
+                payment.partner_id
+                and (payment.customer_wallet_balance - payment.amount)
+                < payment.journal_id.minimum_wallet_amount
+            ):
+                raise UserError(
```
Comment (carmenbianca) [inline account_customer_wallet/models/account_payment.py:33]: In manual testing, this check works, BUT it doesn't test whether any of the move lines actually credit the customer wallet account.

**Review by carmenbianca — COMMENTED** [review]
I mostly succeeded porting this, except one test.

@legalsylvain You're the author of the code I'm struggling to port. Do you have any insights into this? 'No' is a valid answer :)

File: account_customer_wallet/tests/common.py (hunk 1 above)
Comment (legalsylvain) [reply account_customer_wallet/tests/common.py:127]: I just tested.
In V16, create a payment, by default, create an account move line with the account defined globally on the company in the field "Outstanding Receipts Account".
1) I don't know why.
2) I don't know how to handle correctly the payment in V16.

Comment (legalsylvain) [reply account_customer_wallet/tests/common.py:127]: If we set the wallet account on the journal on that field, it works.

image

Comment (carmenbianca) [reply account_customer_wallet/tests/common.py:127]: Thank you Sylvain for that insight. I changed the code to make this work. Commits [66b4a5a](https://github.com/coopiteasy/addons/pull/290/commits/66b4a5a73efcd956e93d30726942878a6a392cb0) and [286a6e2](https://github.com/coopiteasy/addons/pull/290/commits/286a6e2de5ee78a90750fd2e91a05baa8ce998ac) should both be interesting to you. Some of that work may be backported to v12, maybe possibly.

**Comment by legalsylvain** [comment]
@carmenbianca : could you take a time to take a look on this one ? https://github.com/coopiteasy/addons/pull/291

it is now conflicting, because of recent change, but I can rebase, if you think this PR is interesting. (IMO, yes ;-), but let me know).

File: account_customer_wallet/models/account_payment.py [hunk 3]
```diff
@@ -25,6 +25,9 @@ def action_post(self):
         )
 
         for payment in wallet_payments:
+            # TODO: Figure out why/whether this statement is needed. At time of
+            # writing, the tests do not pass without it.
+            payment.partner_id._compute_customer_wallet_balance()
```
Comment (carmenbianca) [inline account_customer_wallet/models/account_payment.py]: I'm unhappy about this.

File: account_customer_wallet/tests/common.py [hunk 4]
```diff
@@ -79,6 +79,9 @@ def _create_move(self, debit=0, credit=0, partner=None):
                 ],
             }
         )
+        move.action_post()
+        partner._compute_customer_wallet_balance()
```
Comment (carmenbianca) [inline account_customer_wallet/tests/common.py]: I'm also unhappy about this.

File: account_customer_wallet/tests/test_balance.py [hunk 5]
```diff
@@ -68,8 +68,9 @@ def test_no_wallet_account(self):
         self._create_move(credit=100)
 
         self.company_id.customer_wallet_account_id = None
-
         self.assertFalse(self.company_id.is_enabled_customer_wallet)
+
+        self.partner._compute_customer_wallet_balance()
```
Comment (carmenbianca) [inline account_customer_wallet/tests/test_balance.py]: And this.

File: account_customer_wallet/models/account_payment.py (hunk 3 above)
Comment (legalsylvain) [reply account_customer_wallet/models/account_payment.py]: That's weird.

Comment (legalsylvain) [reply account_customer_wallet/models/account_payment.py]: Do you want I investigate a little on that topic ?

Comment (carmenbianca) [reply account_customer_wallet/models/account_payment.py]: If you're happy to do so. I'm leaving this PR alone for a little bit; other priorities this week.

Comment (legalsylvain) [reply account_customer_wallet/models/account_payment.py]: "Happy" is a bit of a strong word ;-), but if I can help port this module that GRAP uses...

**Comment by legalsylvain** [comment]
nice !

**Pull request merged**